algorithmic_efficiency/logging_utils.py

```diff
@@ -0,0 +1,492 @@
+        tuning_search_space = json.load(search_space_file)
+        metadata['tuning_search_space'] = tuning_search_space
+
+    workload_properties = _get_workload_properties(self.workload)
+    metadata.update(workload_properties)
+
+    if FLAGS.extra_metadata:
+      extra_metadata = _get_extra_metadata_as_dict(FLAGS.extra_metadata)
+      metadata.update(extra_metadata)
+
+    # System Information
+    metadata['os_platform'] = \
+        platform.platform()  # Ex. 'Linux-5.4.48-x86_64-with-glibc2.29'
+    metadata['python_version'] = platform.python_version()  # Ex. '3.8.10'
+    metadata['python_compiler'] = platform.python_compiler()  # Ex. 'GCC 9.3.0'
+    # Note: do not store hostname as that may be sensitive
+
+    try:
+      metadata['git_branch'] = _get_git_branch()
+      metadata['git_commit_hash'] = _get_git_commit_hash()
+      # Note: do not store git repo url as it may be sensitive or contain a
+      # secret.
+    except:
+      logging.warn('Unable to record git information. Continuing without it.')
+
+    try:
+      metadata['cpu_model_name'] = _get_cpu_model_name()
+      metadata['cpu_count'] = psutil.cpu_count()
+    except:
+      logging.warn('Unable to record cpu information. Continuing without it.')
+
+    gpus = GPUtil.getGPUs()
+    if gpus:
+      try:
+        metadata['gpu_model_name'] = gpus[0].name
+        metadata['gpu_count'] = len(gpus)
+        metadata['gpu_driver'] = gpus[0].driver
+      except:
+        logging.warn('Unable to record gpu information. Continuing without it.')
+
+    # Save workload metadata.json
+    os.makedirs(self.workload_log_dir, exist_ok=True)
+    metadata_filepath = os.path.join(self.workload_log_dir, 'metadata.json')
+    with open(metadata_filepath, 'w', encoding='utf-8') as f:
+      json.dump(metadata, f, ensure_ascii=False, indent=4)
+
+  def _write_package_list_file(self):
+    """Write "packages.txt" to disk.
+
+    It is created at the start of a workload and includes a list of the
+    currently installed OS and python packages."""
+    # Get package lists
+    try:
+      os_package_list = _get_os_package_list()
+      pip_package_list = _get_pip_package_list()
+    except:
+      logging.warn(
+          'Unable to record package information. Continuing without it.')
+      return
+
+    # Save to disk
+    packages_filepath = os.path.join(self.workload_log_dir, 'packages.txt')
+    with open(packages_filepath, 'w', encoding='utf-8') as f:
+      f.write('Python Packages:\n')
+      f.write(pip_package_list)
+      f.write('\n\nOS Packages:\n')
+      f.write(os_package_list)
+
+  def _write_trial_metadata_file(
+      self, workload: spec.Workload,
+      hyperparameters: Optional[spec.Hyperparamters], trial_idx: int,
+      global_step: int, batch_size: int, latest_eval_result: dict,
+      global_start_time: float, accumulated_submission_time: float,
+      goal_reached: bool, is_time_remaining: bool, training_complete: bool,
+      early_stop: bool):
```

Comment:
I know writing it out it is long/annoying, but when defining functions I think it is much cleaner to have a single function parameter per line, instead of a variable number of them per line. that way when scanning argument lists to see if you missed anything, it is easier to match up the orderings. if the linter does not like that, let me know and I'll figure out a way to make it stop.

Reply:
Done. Nice touch! I found a trick to make yapf respect this preference: put a trailing comma after the last arugment, for example:
```
def foo(a, b,):
```

Becomes:
```
def foo(
    a,
    b,
):
```

Tested to work with yapf default settings:
```
$ yapf -i algorithmic_efficiency/logging_utils.py
```